### PR TITLE
[Feature] Enable noisy sims in pulser backend

### LINKED ISSUE #445: [Feature] Enable noisy sims in Pulser backend
Enable noisy expectations in Pulser backend by adding functionality to compute expectations using density matrices in `PulserObservable` forward method.

### PR DESCRIPTION
**Pull request by Roland-djee**
Closes #445

### CODE REVIEW

**Comment by Roland-djee** [comment]
@rajaiitp @dominikandreasseitz Is this OK with you ? If so, I'd like to merge and release to fix protocols.

File: qadence/backends/pulser/convert_ops.py [hunk 1]
```diff
@@ -39,13 +40,14 @@ def forward(
...
-        if self.block.is_parametric:
-            block_mat = block_to_tensor(
-                self.block, values, qubit_support=qubit_support, endianness=endianness
-            ).squeeze(0)
-        else:
-            block_mat = self.block_mat
-
-        return torch.sum(torch.matmul(state, block_mat) * state.conj(), dim=1)
+        # FIXME: cache this, it is very inefficient for non-parametric observables
+        block_mat = block_to_tensor(
```
Comment (dominikandreasseitz) [inline qadence/backends/pulser/convert_ops.py:47]: ok so we cant cache here?

File: qadence/backends/pulser/backend.py [hunk 2]
```diff
@@ -350,10 +353,18 @@ def expectation(
                 # TODO: There should be a better check for batched density matrices.
                 if dms.size()[0] > 1:
                     res_list = [
-                        [obs.native(dm, param_values, qubit_support=support) for dm in dms]
+                        [
+                            obs.native(
+                                dm.squeeze(), param_values, qubit_support=support, noise=noise
```
Comment (dominikandreasseitz) [inline qadence/backends/pulser/backend.py:358]: so we assume dim `0` is always 1? if not, that can result in undefined behaviour

Comment (Roland-djee) [reply qadence/backends/pulser/backend.py:358]: Yes. At this point, `dms` are batched density matrices by the noise parameters. So `dm` is a single density matrix.

File: qadence/backends/pulser/convert_ops.py (hunk 1 above)
Comment (Roland-djee) [reply qadence/backends/pulser/convert_ops.py:47]: I haven't tried to fix this, it predates this PR. Not sure though how I could cache the tensor. Any suggestions ?

Comment (dominikandreasseitz) [reply qadence/backends/pulser/convert_ops.py:47]: the way it was before, which is a very recent MR of mine, did exactly that. in the constructor you check if its a parametric block, if not, you create the tensor and store it as a buffer. in the forward you check if the block is parametric and if not, use the buffer, otherwise build the tensor

File: qadence/backends/pulser/backend.py (hunk 2 above)
Comment (dominikandreasseitz) [reply qadence/backends/pulser/backend.py:358]: ah ok

File: qadence/backends/pulser/convert_ops.py (hunk 1 above)
Comment (Roland-djee) [reply qadence/backends/pulser/convert_ops.py:47]: Alright, lemme have a look.

Comment (Roland-djee) [reply qadence/backends/pulser/convert_ops.py]: Should be in 0c261c2.

**Review by rajaiitp — APPROVED** [review]
looks good to me
